### PR TITLE
SLM-230: add API to verify one time code

### PR DESCRIPTION
**Pull request by nathanrussell-moj-digital**
This PR adds the `POST /oneTimeCode/verify` endpoint to verify a previously created one time code

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/onetimecode/OneTimeCodeService.kt [hunk 1]
```diff
@@ -1,17 +1,42 @@
...
+  private val smokeTestConfig: SmokeTestConfig,
 ) {
 
   fun createAndSendOneTimeCode(email: String, sessionId: String) {
     OneTimeCode(sessionId, oneTimeCodeGenerator.generateOneTimeCode(), email)
       .also { oneTimeCode -> oneTimeCodeRepository.save(oneTimeCode) }
       .also { oneTimeCode -> oneTimeCodeEmailSender.send(email, oneTimeCode.code) }
   }
+
+  fun verifyOneTimeCode(code: String, sessionId: String): String =
```
Comment (nathanrussell-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/onetimecode/OneTimeCodeService.kt:25]: In a previous PR (in fact its the method above), we added an endpoint and service method to create a one time code. Basically a OneTimeCode is an object containing the browser session id (effectively the PK), the one time code, and the users email address. This is stored in redis, and an email with the code sent to the user.

To verify the one time code the browser sends a POST request with the browser session id and the code. The verify service method (ie. this), retrieves the OneTimeCode from the redis repository by session ID, checks the code, then creates the JWT for the user and returns it.
If the redis cache does not contain the OneTimeCode keyed by session ID, an exception is thrown (returned to the client as a 404)

If the OneTimeCode is found but the code does not match, this is also (currently) returned as 404 via the same exception, but the OneTimeCode is not deleted from redis. This is to support the user to retry (ie. they mistyped the code) and to support a future story implementing a '3 strikes and out' type approach.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/onetimecode/OneTimeCodeService.kt [hunk 2]
```diff
@@ -1,17 +1,42 @@
...
+import uk.gov.justice.digital.hmpps.sendlegalmailtoprisonsapi.security.SmokeTestConfig
 
 @Service
 class OneTimeCodeService(
   private val oneTimeCodeGenerator: OneTimeCodeGenerator,
   private val oneTimeCodeRepository: OneTimeCodeRepository,
   private val oneTimeCodeEmailSender: OneTimeCodeEmailSender,
+  private val jwtService: JwtService,
+  private val cjsmService: CjsmService,
+  private val smokeTestConfig: SmokeTestConfig,
```
Comment (joelstobart-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/onetimecode/OneTimeCodeService.kt:16]: @nathanrussell-moj-digital  - It's a bit strange to have a test config on a non-test file?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/onetimecode/OneTimeCodeService.kt [hunk 3]
```diff
@@ -1,17 +1,42 @@
...
 ) {
 
   fun createAndSendOneTimeCode(email: String, sessionId: String) {
     OneTimeCode(sessionId, oneTimeCodeGenerator.generateOneTimeCode(), email)
       .also { oneTimeCode -> oneTimeCodeRepository.save(oneTimeCode) }
       .also { oneTimeCode -> oneTimeCodeEmailSender.send(email, oneTimeCode.code) }
   }
+
+  fun verifyOneTimeCode(code: String, sessionId: String): String =
+    if (code == smokeTestConfig.lsjSecret) {
```
Comment (joelstobart-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/onetimecode/OneTimeCodeService.kt:26]: This doesn't feel very secure, does this mean there is a magic version that disables some security features?

**Review by joelstobart-moj — APPROVED** [review]
LGTM
Acknowledging accepted security consideration

File: src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/onetimecode/OneTimeCodeService.kt (hunk 3 above)
Comment (nathanrussell-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/onetimecode/OneTimeCodeService.kt:26]: Kind of, yes. The smoke test secret is a k8s secret. We have the secret defined in dev and preprod, but not prod.
For this to be a risk in prod (and yes, I accept that the mere fact we have that code is a risk!) we would need to have defined the secret in prod and a bad actor would need to know what that secret was.
Not ideal, but for our smoke tests we need to be able to 'authenticate' our smoke test user without it having a CJSM email address and without it being able to receive and read email

**Pull request merged**